scripts/list_apis.py

```diff
@@ -0,0 +1,277 @@
+
+                qualified_name = f"{module.__name__}.{submodule_name}"
+                try:
+                    submodules.append(importlib.import_module(qualified_name))
+                except Exception as e:
+                    self.errors.append(Crawler.Error(reason=str(e), path=qualified_name))
+
+        return submodules
+
+    def get_module_attributes(self, module: ModuleType) -> Set[str]:
+        """
+        Return the combined list of attributes from dir and __all__
+        """
+        attrs = []
+
+        try:
+            attrs += dir(module)
+            attrs += getattr(module, "__all__", [])
+        except Exception as e:
+            self.errors.append(Crawler.Error(reason=str(e), path=""))
+
+        return set(attrs)
+
+    def crawl(
+        self,
+        obj: Any,
+        name: str,
+        path: List[Item],
+    ) -> None:
+        if self.should_skip(obj, path):
+            return
+
+        # Keep a reference around to ensure id() calls are valid
+        self.all_objects.append(obj)
+
+        if id(obj) in self.apis:
+            # Some things are listed twice, usually as a side-effect of importing
+            # it directly while still in the library. This uses a heuristic of
+            # shorter-path-is-better to pick the best one
+            prev_len = len(self.apis[id(obj)])
+            curr_len = len(path) + 1
+            if curr_len >= prev_len:
+                # It's a longer path and this already has an entry, so don't
+                # both with it any further
+                return
+
+        self.apis[id(obj)] = path + [Crawler.Item(name=name, obj=obj)]
+
+        # Drop down into modules and classes
+        if isinstance(obj, ModuleType):
+            # Explicit crawl through submodules that aren't directly reachable
+            # (e.g. torch.fx) from the parent
+            submodules = self.get_submodules(obj)
+            for submodule in submodules:
+                self.crawl(submodule, submodule.__name__.split(".")[-1], path + [Crawler.Item(name=name, obj=obj)])
+
+            for attr in self.get_module_attributes(obj):
+                item = Crawler.Item(name=name, obj=obj)
+                attr_path = path + [item]
+                try:
+                    next_obj = getattr(obj, attr)
+                except (ModuleNotFoundError, AttributeError) as e:
+                    self.errors.append(Crawler.Error(path=attr_path, reason=str(e)))
+
+                self.crawl(next_obj, attr, attr_path)
+
+    def get_name(self, path: List[str], name: str) -> str:
+        names = [item.name for item in path]
+        return ".".join(names + [name])
+
+    def should_skip(self, obj: Any, path: List[Item]) -> bool:
+        """
+        Return whether 'obj' is relevant to 'relevant_module'. If 'obj' is not a
+        module, then get its module from the __module__ attribute. Then check if
+        the module is part of the relevant_module by comparing their __file__ paths.
```

Comment:
what is `relevant_module`?

Reply:
forgot to update the comment after OOP-ing everything, it's the main module that's being tracked (so this should be `relevant_module` -> `self.module`)